waspc/e2e-test/Main.hs

```diff
@@ -0,0 +1,54 @@
+import Control.Monad (unless)
+import System.Directory (getCurrentDirectory)
+import System.FilePath (joinPath, takeFileName, (</>))
+import System.Info
+import Test.Tasty (TestTree, defaultMain, testGroup)
+import Util
+  ( runGoldenTest,
+    waspCliNew,
+    waspCliNewAndCompile,
+    waspCliNewAndCompileAndMigrate,
+  )
+
+-- TODO: Put `cwd` into ReaderT Monad?
+
+-- TODO: Switch over from FilePath to StrongPath eventually and use helper like so:
+--
+-- data ProjectRoot
+-- getProjectRootPath :: IO (Path Abs (Dir ProjectRoot))
+-- getProjectRootPath = getCurrentDirectory
+
+type GoldenTest = FilePath -> IO TestTree
+
+main :: IO ()
+main = do
+  if os == "mingw32"
+    then putStrLn "Skipping end-to-end tests on Windows due to tests using *nix-only commands"
+    else do
+      -- Stack/Cabal launch `stack test` from root of project, so this should always be some absolute path to waspc
+      cwd <- getCurrentDirectory
+      unless (takeFileName cwd == "waspc") (error "Expecting test process to be invoked from waspc dir")
+      defaultMain =<< tests cwd
```

Comment:
Since this is already in a do block, can't you just do:

```
tests cwd
defaultMain
```
?

Reply:
Hijacking the answer, but actually the equivalent would be
```hs
x <- tests cwd
defaultMain x
```

`=<<` forward the result of the action on the right to the action on the left.

If it was `<<` instead, then what you proposed would be valid replacement, since result of `tests` would be discarded.